### PR TITLE
Assignment2: Kevin + Daniel

### PR DESCRIPTION
**Pull request by kflannery2001**
Partnered w/ @dpalmer10.

### CODE REVIEW

File: Logger/LogFactory.cs [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+
+        //If no file path is configured, return null
+        if (string.IsNullOrWhiteSpace(_filePath))
+        {
+            return null;
+        }
 
-        return null;
+        return new FileLogger(_filePath) { ClassName = className };
     }
```
Comment (Pbrown34) [inline Logger/LogFactory.cs:26]: There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔
The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ✔
All of the above should be unit tested.
Extra Credit
Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses System.Console or System.Diagnostics.Trace ❌

File: Logger/FileLogger.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+namespace Logger
+{
+    public class FileLogger(string filePath) : BaseLogger
+    {
+        private string FilePath { get; } = filePath;
+
+        public override void Log(LogLevel logLevel, string message)
+        {
+            //Format ex: "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
+            string log = $"{DateTime.Now} {ClassName} {logLevel}: {message}{Environment.NewLine}";
```
Comment (Pbrown34) [inline Logger/FileLogger.cs:15]: Your code looks great, I really liked the use of the '!' indicator in your null tests while expecting the error. I may use that method myself next assignment. I couldnt find anything that needed changing so good job there but if you would like to, you could specify how you want your DateTime.Now to be outputted into the file. for example "yyyy-MM-dd HH:mm:ss.fff".

Comment (kflannery2001) [reply Logger/FileLogger.cs:15]: We appreciate the feedback. That's a good tip for formatting DateTime.Now, I'm sure that will be useful in the future.

File: Logger/BaseLoggerMixins.cs [hunk 3]
```diff
@@ -1,6 +1,31 @@
...
+﻿using System;
+using System.Globalization;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
+    public static void Debug(this BaseLogger logger, string message, params object[] args)
+    {
+        ArgumentNullException.ThrowIfNull(logger);
```
Comment (BDullanty) [inline Logger/BaseLoggerMixins.cs:10]: If you refractor these two lines into a method, you can increase maintainability and readability perhaps.

Ex:

public static void CheckNullThenLog(BaseLogger logger, LogLevel loglevel, string message,object[] args)
{
        ArgumentNullException.ThrowIfNull(logger);
        logger.Log(LogLevel.Error, string.Format(CultureInfo.CurrentCulture, message, args));
}

And then just call only one line per log type:

CheckNullThenLog(logger, Loglevel.PickItPerMethod, message, args);

Not a big deal but if you add in more types later this is easier to call, and if you need to swap out something with the actual logic within these method in the future it allows for you to do so only once and still update each type of log.

File: Logger.Tests/FileLoggerTests.cs [hunk 4]
```diff
@@ -1,9 +1,43 @@
...
+        Assert.IsNotNull(logger);
+    }
+
+    [TestMethod]
+    public void FileLogger_LogFileAppended_Success()
+    {
+        // Arrange
+        string message = "Testing123";
+        FileLogger logger = new(_fileName);
+        logger.ClassName = "Test";
```
[the remark this answers is not in the document]
Comment (BDullanty) [reply Logger.Tests/FileLoggerTests.cs:28]: Also perhaps do nameof(FileLoggerTests) here instead of a flat name

**Review by BDullanty — COMMENTED** [review]
- There is an existing `BaseLogger` class. It needs an **auto property** to hold a class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ❌
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
- Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
- Ensure that you turn on EnforceCodeStyleInBuild ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
- Turn on Nullability (`Nullable`) ✔
- **All of the above should be unit tested.**

## Extra Credit

- Implement an additional logger. This logger must be unit tested. Some options to consider could be one that uses `System.Console` or `System.Diagnostics.Trace` ❌
- Implement the factory pattern using static abstract methods on a Logger interface instead. ❌

File: Logger/FileLogger.cs (hunk 2 above)
Comment (dpalmer10) [reply Logger/FileLogger.cs:15]: Thank you! Also I think that could be a good small addition, perhaps using UTC to try and standardize the logger to be used across multiple time zones without confusion.

File: Logger.Tests/LogFactoryTests.cs [hunk 5]
```diff
@@ -1,9 +1,57 @@
...
+    }
+
+    [TestMethod]
+    public void CreateLogger_CreateValidLogger_Success()
+    {
+        // Arrange
+        LogFactory factory = new();
+        factory.ConfigureFilePath("test.txt");
+        // Act
+        var logger = factory.CreateLogger("Test");
```
[the remark this answers is not in the document]
Comment (dpalmer10) [reply Logger.Tests/LogFactoryTests.cs:28]: good catch, thanks.

File: Logger.Tests/FileLoggerTests.cs [hunk 6]
```diff
@@ -1,9 +1,43 @@
...
+    {
+        // Arrange
+        FileLogger logger = new(_fileName);
+        // Act
+        // Assert
+        Assert.IsNotNull(logger);
+    }
+
+    [TestMethod]
+    public void FileLogger_LogFileAppended_Success()
```
[the remark this answers is not in the document]
Comment (kflannery2001) [reply Logger.Tests/FileLoggerTests.cs:23]: Thank you, that is very helpful.

File: Logger/FileLogger.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+            //Print log message to file
```
[the remark this answers is not in the document]
Comment (dpalmer10) [reply Logger/FileLogger.cs:16]: good idea, looking closer at the CreateDirectory method, the if statement isn't needed as the line will be skipped over if the directory already exists.

File: Logger/LogFactory.cs [hunk 8]
```diff
@@ -1,10 +1,27 @@
-﻿namespace Logger;
+﻿using System;
+
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _filePath;
```
[the remark this answers is not in the document]
Comment (kflannery2001) [reply Logger/LogFactory.cs]: Fixed

File: Logger/BaseLoggerMixins.cs (hunk 3 above)
Comment (kflannery2001) [reply Logger/BaseLoggerMixins.cs:10]: Very good catch. Moving forward, we will take that concept into account to increase maintainability.

File: Logger.Tests/FileLoggerTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kflannery2001) [reply Logger.Tests/FileLoggerTests.cs]: Good ideas, we've updated accordingly.

File: Logger/FileLogger.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (BDullanty) [reply Logger/FileLogger.cs:16]: Oh that's cool, thanks for the tip!

File: Logger/LogFactory.cs [hunk 9]
```diff
@@ -1,10 +1,27 @@
...
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? FilePath {  get; set; }
+
+    public string ConfigureFilePath(string filePath)
     {
+        ArgumentNullException.ThrowIfNull(filePath);
+        FilePath = filePath;
+        return FilePath;
```
[the remark this answers is not in the document]
Comment (OliviaJSmith) [reply Logger/LogFactory.cs:13]: although if you only use this value for a test probably need to reconsider your test

**Pull request closed without merging**